src/type/validate.js

```diff
@@ -593,6 +759,143 @@ function createInputObjectCircularRefsValidator(
+    inputObj: GraphQLInputObjectType,
+  ): void {
+    // Start with an empty object as a way to visit every field in this input
+    // object type and apply every default value.
+    return detectValueDefaultValueCycle(inputObj, {});
+  };
+
+  function detectValueDefaultValueCycle(
+    inputObj: GraphQLInputObjectType,
+    defaultValue: mixed,
+  ): void {
+    // If the value is a List, recursively check each entry for a cycle.
+    // Otherwise, only object values can contain a cycle.
+    if (isIterableObject(defaultValue)) {
+      for (const itemValue of defaultValue) {
+        detectValueDefaultValueCycle(inputObj, itemValue);
+      }
+      return;
+    } else if (!isObjectLike(defaultValue)) {
+      return;
+    }
+
+    // Check each defined field for a cycle.
+    for (const field of Object.values(inputObj.getFields())) {
+      const fieldType = getNamedType(field.type);
+
+      // Only input object type fields can result in a cycle.
+      if (!isInputObjectType(fieldType)) {
+        continue;
+      }
+
+      // If the provided value has this field defined, recursively check it for
+      // cycles instead of checking the field's default value.
+      if (hasOwnProperty(defaultValue, field.name)) {
+        detectValueDefaultValueCycle(fieldType, defaultValue[field.name]);
+        continue;
+      }
+
+      detectFieldDefaultValueCycle(inputObj, field, fieldType);
+    }
+  }
+
+  function detectLiteralDefaultValueCycle(
+    inputObj: GraphQLInputObjectType,
+    defaultValue: ConstValueNode,
+  ): void {
+    // If the value is a List, recursively check each entry for a cycle.
+    // Otherwise, only object values can contain a cycle.
+    if (defaultValue.kind === Kind.LIST) {
+      for (const itemLiteral of defaultValue.values) {
+        detectLiteralDefaultValueCycle(inputObj, itemLiteral);
+      }
+      return;
+    } else if (defaultValue.kind !== Kind.OBJECT) {
+      return;
+    }
+
+    // Check each defined field for a cycle.
+    const fieldNodes = keyMap(defaultValue.fields, (field) => field.name.value);
+    for (const field of Object.values(inputObj.getFields())) {
+      const fieldType = getNamedType(field.type);
+
+      // Only input object type fields can result in a cycle.
+      if (!isInputObjectType(fieldType)) {
+        continue;
+      }
+
+      // If the provided value has this field defined, recursively check it for
+      // cycles instead of checking the field's default value.
+      if (hasOwnProperty(fieldNodes, field.name)) {
+        detectLiteralDefaultValueCycle(fieldType, fieldNodes[field.name].value);
+        continue;
+      }
+
+      detectFieldDefaultValueCycle(inputObj, field, fieldType);
+    }
+  }
+
+  function detectFieldDefaultValueCycle(
+    inputObj: GraphQLInputObjectType,
+    field: GraphQLInputField,
+    fieldType: GraphQLInputObjectType,
+  ): void {
+    // Only a field with a default value can result in a cycle
```

Comment:
We below on falsyness below, but `0`, `""`, etc are valid defaultValues that are falsy. We should be clear that only lists/objects can cause cycles and that's why we're using the simple falsy check.

```suggestion
    // Only a field with a default value that's a list or object can result in a cycle
```

Reply:
The falsey check is on a `GraphQLDefaultValueUsage` object, so it's either that object or `undefined` - so this check is actually looking to see if a default value is defined at all

Reply:
Ah yes [it's a GraphQLDefaultValueUsage](https://github.com/graphql/graphql-js/blob/fceda32845ae9950c926cfe6b05af403da63ad60/src/type/definition.ts#L1778) :+1: